### DISCUSSION TITLE
Session query exception, is it necessary to rollback session?

### DISCUSSION THREAD

**Discussion by heilutan** [sqlalchemy/sqlalchemy#11945] [Usage Questions] [answered]
```python
engine = sqlalchemy.create_engine(connect_url, pool_size=10, max_overflow=20, echo=False, pool_recycle=300, pool_pre_ping=True, pool_use_lifo=True)
Session = sessionmaker(bind=engine)

def get_session():
    return Session
```

Example1

```python
def query_func():
    session = get_session()
    try:
        session.query()
    except:
        session.rollback()  # Is it necessary to rollback session?
```

Q1: This method does not start a transaction by default, so is it okay to continue using the session for subsequent queries, updates, or deletes?

Example2

```python
def loop_func():
    session = get_session()
    while True:
        try:
            session.query()
        except:
           session.rollback() #Is it necessary to rollback session? The error "psycopg2.OperationalError: SSL connection has been closed unexpectedly" occurs at regular intervals.
```

Q2: In `loop_func`, is it necessary to rollback session for subsequent queries, updates, or deletes?

**Comment by gordthompson** [comment]
Have you seen this?

[When do I construct a Session, when do I commit it, and when do I close it?](https://docs.sqlalchemy.org/en/20/orm/session_basics.html#when-do-i-construct-a-session-when-do-i-commit-it-and-when-do-i-close-it)

**Reply by heilutan** [reply]
Hi gordthompson,

I notice this link. The Session, whenever it is used to talk to the database, begins a database transaction as soon as it starts communicating. I understand that when an exception occurs during a session query, it may cause the session's state or transaction to become invalid. Therefore, I assumed that a `session.rollback` is necessary before performing the next session operation (query, update, delete). However, in actual use, I noticed that even without calling `session.rollback`, no issues arose.

**Reply by heilutan** [reply]
Is it because `query` only performs read operations, so there are no data consistency issues, and that's why it doesn't cause problems even without a rollback?

**Reply by CaselIT** [reply]
Rollback is only used when there are errors. If there are no errors you don't need to call rollback.

**Reply by heilutan** [reply]
So, is a rollback necessary when an exception occurs during a query?

**Comment by heilutan** [comment]
`def add_user(username, email):`
`    try:`
`        new_user = User(username=username, email=email)`
`        db.session.add(new_user)`
`        db.session.commit()`
`        return {'message': 'User added successfully!'}`
`    except Exception as e:`
`        db.session.rollback()  # Rollback in case of error`
`        return {'error': str(e)}, 500`
`    finally:`
`        db.session.close()  # Make sure to close the session`

If an exception occurs during a session `add` or `update`, it will cause transaction inconsistency, so a rollback is necessary. However, since a session `query` only performs read operations, is a rollback also necessary if an exception occurs?

**Reply by CaselIT** [reply, marked as answer]
If you want to continue using the session yes.
Note that closing the session will also call rollback so it's not needed in that case

**Reply by heilutan** [reply]
`def query_func():`
`    session = get_session()`
`    try:`
`        session.query()`
`    except:`
`        session.rollback()  # Is it necessary to rollback session?`

Why is a rollback necessary if you want to continue using the session after a query exception? My colleague believes that since `query` doesn't modify the database, a rollback is an unnecessary operation.

**Comment by heilutan** [comment]
Is it necessary to catch exceptions and roll back for all `session.query` operations? Do you have a best practice example for using `query`?
